src/Lean/Meta/Closure.lean

```diff
@@ -203,9 +203,21 @@ partial def collectExprAux (e : Expr) : ClosureM Expr := do
     let type ← collect type
     let newFVarId ← mkFreshFVarId
     let userName ← mkNextUserName
+    /-
+    If the metavariable is a delayed assignment, it would be an error to push it onto `exprMVarArgs` directly,
+    since the assumption is that delayed assignment metavariables are always fully applied (for example `instantiateMVars` assumes this).
+    The need to handle delayed assignment metavariables was uncovered in issue https://github.com/leanprover/lean4/issues/6354
+    -/
+    let e' ←
+      if let some { fvars, .. } ← getDelayedMVarAssignment? mvarId then
+        -- Eta expand `e` for the requisite number of arguments.
+        forallBoundedTelescope mvarDecl.type fvars.size fun args _ => do
+          mkLambdaFVars args <| mkAppN e args
```

Comment:
Have you considered other solutions that do not require eta-expansion for this issue, even if they apply only to special cases? If they are not worth pursuing, we should document that here.

Reply:
> It did cross my mind that I could try matching on the case where the delayed assignment metavariable is applied to fvars, but it wasn't clear to me that the added code complexity would be worth it. We would still need the eta expansion code path for correctness.

Let's add an extra comment documenting this decision.

Reply:
Added.

(I also tried to implement this suggestion, but I couldn't get it to work within the time I gave toward attempting it. Here's [the branch](https://github.com/leanprover/lean4/compare/master...kmill:fix_6354_fvars) with non-working code.)

Reply:
Thanks!